main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -545,61 +570,61 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
           val shadowedVar = shadowing(sym, env0)
 
           // Visit the channel and body expressions.
-          val chanUsed = visitExp(chan, env1)
-          val bodyUsed = visitExp(body, env1)
+          val chanUsed = visitExp(chan, env1.resetApplies)
+          val bodyUsed = visitExp(body, env1.resetApplies)
 
           // Check if the variable symbol is dead in the body.
           if (deadVarSym(sym, bodyUsed))
-            (chanUsed ++ bodyUsed ++ shadowedVar) - sym + UnusedVarSym(sym)
+            (chanUsed and bodyUsed and shadowedVar) - sym + UnusedVarSym(sym)
           else
-            (chanUsed ++ bodyUsed ++ shadowedVar) - sym
+            (chanUsed and bodyUsed and shadowedVar) - sym
       }
 
-      rulesUsed.foldLeft(Used.empty ++ defaultUsed) {
-        case (acc, used) => acc ++ used
+      rulesUsed.foldLeft(defaultUsed) {
+        case (acc, used) => acc or used
       }
 
-    case Expression.ProcessSpawn(exp, _, _, _) => Used.empty ++ visitExp(exp, env0)
+    case Expression.ProcessSpawn(exp, _, _, _) => visitExp(exp, env0.resetApplies)
 
-    case Expression.ProcessSleep(exp, _, _, _) => Used.empty ++ visitExp(exp, env0)
+    case Expression.ProcessSleep(exp, _, _, _) => visitExp(exp, env0.resetApplies)
 
     case Expression.ProcessPanic(msg, _, _, _) => Used.empty
 
     case Expression.FixpointConstraintSet(cs, _, _, _) =>
       cs.foldLeft(Used.empty) {
-        case (used, con) => used ++ visitConstraint(con, env0)
+        case (used, con) => used and visitConstraint(con, env0.resetApplies)
       }
 
     case Expression.FixpointCompose(exp1, exp2, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      us1 ++ us2
+      val us1 = visitExp(exp1, env0.resetApplies)
+      val us2 = visitExp(exp2, env0.resetApplies)
+      us1 and us2
 
     case Expression.FixpointSolve(exp, _, _, _, _) =>
-      visitExp(exp, env0)
+      visitExp(exp, env0.resetApplies)
 
     case Expression.FixpointProject(sym, exp, _, _, _) =>
-      val us = visitExp(exp, env0)
-      Used.of(sym) ++ us
+      val us = visitExp(exp, env0.resetApplies)
+      Used.of(sym) and us
 
     case Expression.FixpointEntails(exp1, exp2, _, _, _) =>
-      val used1 = visitExp(exp1, env0)
-      val used2 = visitExp(exp2, env0)
-      used1 ++ used2
+      val used1 = visitExp(exp1, env0.resetApplies)
+      val used2 = visitExp(exp2, env0.resetApplies)
+      used1 and used2
 
     case Expression.FixpointFold(sym, exp1, exp2, exp3, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      val us3 = visitExp(exp3, env0)
-      Used.of(sym) ++ us1 ++ us2 ++ us3
+      val us1 = visitExp(exp1, env0.resetApplies)
+      val us2 = visitExp(exp2, env0.resetApplies)
+      val us3 = visitExp(exp3, env0.resetApplies)
```

Comment:
So, one thing that is happening is that we are calling resetApplies over and over. This causes a lot of allocation. Maybe that is the cause of the slow down? For example in this case, we could construct env1 = env0.resetApplies and save at least one allocation.

Reply:
It is not so pretty, but if we make the `applies` part of recursionContext into an ordinary parameter of visitExp then we can avoid a whole lot of allocation.

Reply:
Or alternatively, it seems that applies is reset almost everywhere, so then we could just have a helper function that we call in the Apply case which counts the number of nested Apply's. If that makes sense? I don't know if this will mean that we miss more warnings that previously?

Reply:
I've added an alternative very easy-to-implement optimization to the `resetApplies` method which just checks whether `resetApplies` would return an exact copy, and in that case just return `this`.

It's actually a little hard to tell what is just normal variation and what is actually different run-to-run of the benchmark since there seems to be a lot of variability when running on my machine. Here is the output of two tests on identical source:

```
Reader, 1581045117, 2711000
Parser, 1581045117, 55395300
Weeder, 1581045117, 21127200
Namer, 1581045117, 21813700
Resolver, 1581045117, 50492800
Typer, 1581045117, 536938400
Stratifier, 1581045117, 7407200
PatternExhaustiveness, 1581045117, 4149100
Redundancy, 1581045117, 14266300
Linter, 1581045117, 800
Safety, 1581045117, 2326600
Documentor, 1581045117, 1700
Monomorph, 1581045117, 265885200
Synthesize, 1581045117, 5708400
Simplifier, 1581045117, 53888600
Uncurrier, 1581045117, 10100
ClosureConv, 1581045117, 23709500
LambdaLift, 1581045117, 5942500
Tailrec, 1581045117, 2788400
Optimizer, 1581045117, 5964200
TreeShaker, 1581045117, 30666100
VarNumbering, 1581045117, 3611200
Finalize, 1581045117, 25744400
JvmBackend, 1581045117, 466768500

Reader, 1581045749, 2641600
Parser, 1581045749, 58979900
Weeder, 1581045749, 19216900
Namer, 1581045749, 22574700
Resolver, 1581045749, 45464700
Typer, 1581045749, 510456600
Stratifier, 1581045749, 7681500
PatternExhaustiveness, 1581045749, 4684000
Redundancy, 1581045749, 17039000
Linter, 1581045749, 800
Safety, 1581045749, 2382200
Documentor, 1581045749, 1600
Monomorph, 1581045749, 229490900
Synthesize, 1581045749, 22991900
Simplifier, 1581045749, 47061100
Uncurrier, 1581045749, 11200
ClosureConv, 1581045749, 24361500
LambdaLift, 1581045749, 6066000
Tailrec, 1581045749, 3011800
Optimizer, 1581045749, 5772700
TreeShaker, 1581045749, 31185500
VarNumbering, 1581045749, 3518300
Finalize, 1581045749, 23489100
JvmBackend, 1581045749, 417791300
```